File: STRUKTUR DATA (SMSTR 2)/materi UTS/mandiriUTS.py
"""
# record pada python
#define class
class beverage:
    def __init__(self, nama, harga):
        self.nama = nama
        self.harga = harga
#membuat objek tunggal dari class beverage
p1 = beverage("bajigur", 25000)
#membuat array objek dari class beverage
p = []
p.append(beverage("bajigur", 25000))
p.append(beverage("es doger", 15000))
p.append(beverage("puding", 5000))
p.append(beverage("cendol", 10000))
#cara mengakses
print(p1.nama, p1.harga)
print(p[0].nama, p[0].harga)
print(p[1].nama, p[1].harga)
print(p[2].nama, p[2].harga)
print(p[3].nama, p[3].harga)
"""
"""
#array
m=[[1,2,3], [4,5,6], [7,8,9]]
print(m)
print(m[0])
print(m[2])
print(m[0][0], m[0][1], m[0][2])
print(m[1][0], m[1][1], m[1][2])
for i in m:
    print(i)
for i in m:
    for j in i:
        print(j)
for i in range(0, 3):
    for j in range(0, 3):
        if j < 2:
            print(m[i][j], end=" ")
        else:
            print(m[i][j])

m=[1, 2, 3, 4, 5, 6, 7, 8, 9]
print(m)
print(m[0])
print(m[1])
print("panjang: ", len(m))
for i in range(0, 9):
    print(m[i], end=" ")
print()
for i in m:
    print(i)
m.append(10)
for i in range(0, 10):
    print(m[i], end=" ")
print("panjang: ", len(m))
m.pop(9)
for i in range(0, len(m)):
    print(m[i], end=" ")
print("panjang: ", len(m))
m.pop(5)
for i in range(0, len(m)):
    print(m[i], end = " ")
print("panjang: ", len(m))
print()
m.insert(5, 100)
print(m)
m[3] = 999
print(m)
m.sort()
print(m)
m.reverse()
print(m)
m.remove(2)
print(m)
n = [111, 112, 113]
m.extend(n)
print(m)
m.clear()
print(m)
"""
"""
#pembutan kelas dan objek
#mendefinisikan kelas
class Kendaraan:
    def __init__(self, jenis, roda):
        self.jenis = jenis
        self.roda = roda
    def info(self):
        print(f"jenis: {self.jenis}, roda: {self.roda}")
#membuat objek dari kelas kendaraan
mobil = Kendaraan(jenis="mobil", roda=4)
sepeda =Kendaraan(jenis="sepeda", roda=2)
#mengakses atribut dan method objek
print(mobil.jenis)
sepeda.info()
"""
"""
#pembuatan single linked list view all and add first
class SingleLL:
    #pembuatan kelas node
    class _Node:
        def __init__(self, element, nextNode=None):
            self.element = element
            self.nextNode = nextNode
    #pembuatan single linked list
    def __init__(self):
        self._head = None
        self._size = 0
    #pembuatan fungsi view all
    def __str__(self):
        result = " "
        pointer = self._head
        while pointer != None:
            result = result + str(pointer.element) + " "
            pointer = pointer.nextNode
        return result
    #add first pada ll yang maih kosong
    def add_first(self, element):
        newNode = self._Node(element)
        newNode.nextNode = self._head
        self._head = newNode
        self._size += 1
    def __len__(self):
        return self._size
def main():
    mylist = SingleLL()
    mylist.add_first(50)
    mylist.add_first(30)
    mylist.add_first(10)
    print(str(mylist))
    print(len(mylist))
main()
"""
"""
#contoh insert ditengah

#pembuatan single linked list view all and add first
class SingleLL:
    #pembuatan kelas node
    class _Node:
        def __init__(self, element, nextNode=None):
            self.element = element
            self.nextNode = nextNode
    #pembuatan single linked list
    def __init__(self):
        self._head = None
        self._size = 0
    #pembuatan fungsi view all
    def __str__(self):
        result = " "
        pointer = self._head
        while pointer != None:
            result = result + str(pointer.element) + " "
            pointer = pointer.nextNode
        return result
    #add first pada ll yang maih kosong
    def add_first(self, element):
        newNode = self._Node(element)
        newNode.nextNode = self._head
        self._head = newNode
        self._size += 1
    #menambahkan node ditengah linkedlist
    def add_middle(self, index, element):
        if index < 0 or index > self._size:
            raise IndexError("index out of range")
        if index == 0:
            self.add_first(element)
        else:
            newNode = self._Node(element)
            current_node = self._head
            for _ in range(index - 1):
                current_node = current_node.nextNode
            newNode.nextNode = current_node.nextNode
            current_node.nextNode = newNode
            self._size += 1

    def __len__(self):
        return self._size
def main():
    mylist = SingleLL()
    mylist.add_first(50)
    mylist.add_first(30)
    mylist.add_first(10)
    print("awal")
    print(str(mylist))
    print(len(mylist))
    #menambahkan node
    mylist.add_middle(2, 20)
    print("setelah")
    print(str(mylist))
    print(len(mylist))
main()
"""
"""
#delete pada akhir list
class SingleLL:
    #pembuatan kelas node
    class _Node:
        def __init__(self, element, nextNode=None):
            self.element = element
            self.nextNode = nextNode
    #pembuatan single linked list
    def __init__(self):
        self._head = None
        self._size = 0
    #pembuatan fungsi view all
    def __str__(self):
        result = " "
        pointer = self._head
        while pointer != None:
            result = result + str(pointer.element) + " "
            pointer = pointer.nextNode
        return result
    #add first pada ll yang maih kosong
    def add_first(self, element):
        newNode = self._Node(element)
        newNode.nextNode = self._head
        self._head = newNode
        self._size += 1
    #menghapus node diakhir linked list
    def delete_last(self):
        if self._head is None:
            raise IndexError("linked list os empty")
        elif self._size == 1:
            self._head = None
        else:
            current_node = self._head
            while current_node.nextNode.nextNode is not None:
                current_node = current_node.nextNode
            current_node.nextNode = None
        self._size -=1
    def __len__(self):
        return self._size
def main():
    mylist = SingleLL()
    mylist.add_first(50)
    mylist.add_first(30)
    mylist.add_first(10)
    print("awal: ")
    print(str(mylist))
    print(len(mylist))
    #menghapus node
    mylist.delete_last()
    print("setelah delete last: ")
    print(str(mylist))
    print(len(mylist))
main()
"""

File: STRUKTUR DATA (SMSTR 2)/materi UTS/mandiriUTS.cpp
/*
// MEMBUAT. HEAD 42 57 23 51 NULL
#include <iostream>
using namespace std;
//struct node untuk menyimpab data dan pointe kenode berikutnya
struct Node{
    int data;
    Node* next;
};
//fungsi untuk menambahkan node baru diakhir linkedlist
void appendNode(Node** head, int data){
    Node* newNode = new Node;
    newNode -> data = data;
    newNode -> next = nullptr;
    if (*head == nullptr){
        *head = newNode;
        return;
    }
    Node* current = *head;
    while (current -> next != nullptr){
        current = current -> next;
    }
    current -> next = newNode;
}
//fungsi untuk menampilkan sebuah node dalam linkedlist
void printList(Node* head){
    cout <<"head-> ";
    Node* current = head;
    while (current != nullptr){
        cout << current -> data <<"->";
        current = current -> next;
    }
    cout <<"null"<<endl;
}
int main(){
    Node* head = nullptr;
    //menambahkan node kedalam linkedlist
    appendNode(&head, 42);
    appendNode(&head, 57);
    appendNode(&head, 23);
    appendNode(&head, 51);
    //menampilkan linkedlist
    printList(head);
    return 0;
}
*/

/*
//LIST = [42, 57, 23, 51] TUKAR POSISI LIST BERDASARKAN INPUTAN USER
#include<iostream>
#include<vector>
using namespace std;
int main(){
    //list awal
    vector <int> my_list = {42, 57, 23, 51};
    //menampilkan list saat ini
    cout<<"list saat ini: ";
    for (int x: my_list){
        cout <<x<< " ";
    }
    cout<<endl;
    //meminta input dari pengguna
    int pos1, pos2;
    cout <<"masukkan posisi elemen pertama yang ingin ditukar(0-based):";
    cin>>pos1;
    cout <<"masukkan posisi elemen kedua yang ingin ditukar(0-based):";
    cin>>pos2;
    //validasi input posisi
if (pos1 < 0 || pos1 >= my_list.size() || pos2<0 ||pos2 >= my_list.size()){
    cout <<"posisi yang dimasukkan tidak valid" <<endl;
    return 1;
}
    //menukar posisi element
    swap(my_list[pos1], my_list[pos2]);
    //menampilkan list setelah ditukar
    cout <<"list setelah ditular: ";
    for (int x:my_list){
        cout << x<<" ";
    }
    cout << endl;
    return 0;
}
*/



// MEMODIFIKASI SINGLE LL
#include<iostream>
using namespace std;
//structur node untuk menyimpan data dan pointer ke node berikutnya
struct Node{
    int data;
    Node* next;
};
// fungsi untuk menambahkan node baru diakhir ll
void appendNode(Node** head, int data){
    Node* newNode = new Node;
    newNode -> data = data;
    newNode -> next = nullptr;
    if (*head == nullptr){
        *head = newNode;
        return;
    }
    Node* current = *head;
    while (current -> next != nullptr){
        current = current -> next;
    }
    current -> next = newNode;

}
//fungsi untul menambahkan node baru setelah node tertentu
void insertAfter(Node* prevNode, int data){
    if (prevNode == nullptr){
        cout <<"node sebelumnta tidak ditemukan dalam ll "<<endl;
        return;
    }
    Node* newNode = new Node;
    newNode -> data = data;
    newNode -> next = prevNode -> next;
    prevNode -> next = newNode;
}
//fungsi untuk menyisipkan node baru sebelum node tertentu
void insertBefore(Node** head, Node* nextNode, int data){
    if (nextNode == nullptr){
        cout <<"node selanjutnya tidak ditemukan dalam linkedlist"<<endl;
        return;
    }
    Node* newNode = new Node;
    newNode -> data = data;
    if (*head == nextNode){
        newNode -> next = *head;
        *head = newNode;
        return;
    }
    Node* current = *head;
    while (current-> next != nextNode){
        current = current -> next;
    }
    newNode-> next = current -> next;
    current -> next = newNode;
}
//fungsi untuk menghapus node dengan nnilai tertentu
void deleteNode(Node** head, int data){
    if(*head == nullptr){
        cout <<"linkedlist kosong"<< endl;
        return;
    }
    if ((*head)-> data == data){
        Node* temp = *head;
        *head = (*head)-> next;
        delete temp;
        return;
    }
    Node* current = *head;
    while (current -> next != nullptr){
        if (current -> next -> data == data){
            Node* temp = current -> next;
            current -> next = current -> next -> next;
            delete temp;
            return;
        }
        current = current -> next;
    }
    cout <<"node dengan nilai "<<data<<"tidak ditemukan dalam ll"<<endl;
}
//fungsi untuk menampilkan seluruh node dalam mlinkedlist
void printList(Node* head){
    cout <<"head -> ";
    Node* current = head;
    while (current != nullptr){
        cout << current -> data <<"-> ";
        current=current-> next;
    }
    cout <<"null"<<endl;
}
int main(){
    Node* head = nullptr;
    //menambahkan node kedalam linked list
    appendNode(&head, 42);
    appendNode(&head, 57);
    appendNode(&head, 23);
    appendNode(&head, 51);
    cout<<"linkedlist awal: "<< endl;
    printList(head);
    //menambahkan node 10 setelah node 51
    Node* currentNode = head;
    while (currentNode -> data != 51){
        currentNode = currentNode -> next;
    }
    insertAfter(currentNode, 10);
    //menyisipkan node 10 sebelum node 51
    currentNode = head;
    while (currentNode -> next -> data != 51){
        currentNode = currentNode -> next;
    }
    insertBefore(&head, currentNode -> next, 10);
    //menghapus node 23
    deleteNode(&head, 23);
    cout <<"linkedlist setelah modifikasi: "<<endl;
    printList(head);
    return 0;
}





